stdlib/public/Compatibility50/DynamicReplaceable.cpp

```diff
@@ -0,0 +1,88 @@
+//===--- ProtocolConformance.cpp - Swift protocol conformance checking ----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Runtime support for dynamic replaceable functions.
+//
+// This implementation is intended to be backward-deployed into Swift 5.0
+// runtimes.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Overrides.h"
+#include "swift/Runtime/Exclusivity.h"
+#include "swift/Runtime/Metadata.h"
+#include "../runtime/ThreadLocalStorage.h"
+
+using namespace swift;
+
+// Mirror the SwiftTLSContext, which is defined in the runtime.
+// This is a hack: the layout must match with SwiftTLSContext defined in
+// Exclusivity.cpp. Fortunately it contains only a single pointer (the head of
+// the access list).
+class SwiftTLSContext {
+  void *AccessHead;
+
+public:
+  void *getHead() const { return AccessHead; }
+
+  void setExtraBit() {
+    AccessHead = (void *)((intptr_t)AccessHead | 1);
+  }
+
+  void clearExtraBit() {
+    AccessHead = (void *)((intptr_t)AccessHead & ~1);
+  }
+
+  bool testExtraBit() const {
+    return (bool)((intptr_t)AccessHead & 1);
+  }
+};
+
+__attribute__((visibility("hidden"), weak))
+extern "C" char *swift_getFunctionReplacement(char **ReplFnPtr, char *CurrFn) {
```

Comment:
I'm really wary about baking this into apps. These are pretty simple functions, which means there isn't much potential for breaking changes, but I'm paranoid that something would come up eventually.

Can we find a way to make these call the versions from `libswiftCore.dylib` if they exist? ARCLite accomplishes this using a terrifying technique that involves patching entries in the lazy pointer section at runtime. We could potentially do that, although it is scary.

Another option might be to use `dlsym` in these functions to try to find one in `libswiftCore`, and then call through to it if it exists. That would add a bit more overhead but it wouldn't be too horrible.

We could cut down on that overhead by having the static library include global variable function pointers that point to these functions, similar to how `swift_retain` etc. work. When code is compiled for older targets, have it call through the function pointers. The function pointers can point to something that decides whether to use the version from `libswiftCore` or the version here, then writes the appropriate function pointer back to the global variable.

Reply:
Yes. Andy pointed out that the current implementation would bake in the ABI of how we do things.

Jordan and I talked yesterday and he proposed marking the functions in the libswiftCore.dylib weak. Something like

```
libswiftCore.dylib:

__attribute__(weak) // or is that __attribute__(weak_import)
void *swift_getReplacement();
```
The compatibility archive can then check whether the weak symbols is bound and use it if it is otherwise fall back to a compatibility version. This way you always get the current implementation and only fall back if there is not implementation. The fallback does not become ABI this way.

```
libcompatdynrepl.a:

void *swift_getReplacement50() {
  if (swift_getReplacement)
    return swift_getReplacement()
  ... fallback implementation
}
```

Reply:
Good call, this is like my `dlsym` idea except much better.